Add scroll-spy functionality

rogeruiz opened the pull request:
@maya @juliaelman @yozlet

rogeruiz committed:
Add scroll-spy functionality

maya commented:
Isn't working for research report for me in Chrome when I click on it: 

![](http://g.recordit.co/p7thpyrRVz.gif)

Update: sometimes it works sometimes it doesn't?

maya commented:
@rogeruiz jk just tried again and that issue seems to be fine!

maya commented:
I know @carodew says she tested the Galaxy S5, but these were my results:

![](http://g.recordit.co/BMWHAQrjSh.gif)

Here's the iPhone 6 (everything except overview works on click):

![](http://g.recordit.co/MaBfY8W89F.gif)

Large screens seem :+1: for me.

rogeruiz commented:
Hey @maya Thanks for the screens. I think I cleared up the bugs related to the highlighting. I'll point them out in the line-notes.

rogeruiz commented on assets/scripts/components/scroll-spy.js:
```diff
@@ -0,0 +1,49 @@
+  return [].map.call( $links, function (el) {
+    var $link = $(el);
+    var $section = $($link.attr('href'));
+    var topOffset = $section.position().top - originalNavigationHeight;
```
Make sure that the sticky navigation's height is taken into account for the `topOffset` of all sections.

rogeruiz commented on assets/scripts/components/scroll-spy.js:
```diff
@@ -0,0 +1,49 @@
+
+  var scrollY = $(window).scrollTop();
+
+  var remainingHeight = $(document).height() - scrollY - window.innerHeight;
```
Calculate the remaining height of the page based on the current scroll y-position.

rogeruiz commented on assets/scripts/components/scroll-spy.js:
```diff
@@ -0,0 +1,49 @@
+      section.$link.addClass('is-active');
+    }
+
+    if (remainingHeight < window.innerHeight * 0.25) {
```
If the remaining height of the page is less than 25% of the height of the page, the user will never be able to scroll to meet the `topOffset` calculated above. So, remove all the `is-active` classes and apply `is-active` to the last item in the `sections` array.

rogeruiz replied:
This assumption works for this specific navigation and sections number and isn't suited to other kinds of navigation / section combinations.

rogeruiz commented on assets/styles/homepage.scss:
```diff
@@ -268,6 +272,7 @@ h2 {
 // Contact section extra padding so scrollspy works
 .section-end {
   @include media ($medium-screen) {
+    padding-top: 13rem;
```
I included some extra padding on the top of the Contact section to visually balance the contact section. It was too top heavy before. I'm okay with reverting this change, NBD.

maya replied:
I would not add this, to keep it consistent with the others. Let's change it back.

rogeruiz replied:
Sounds good @maya I'll remove it.

rogeruiz committed:
Update `js-scroll-spy` to be an `id`;

This patch makes the `js-scroll-spy` hook an `id` since there shouldn't
be more than one scroll-spy on a page.

maya commented:
iPhone 6 and Galaxy S5 look good to me!

rogeruiz committed:
Fix rounding error for calculate-anchor-position

rogeruiz committed:
Fix undefined error

maya committed:
Fix firefox scroll bug

rogeruiz commented:
Hey @juliaelman or @msecret. Could one of you take a look at the JS for this PR? @maya and I have tested it in all major browsers. Thanks!

msecret commented:
LGTM

maya commented:
:rocket:
